### PR TITLE
Fix bug when DELETE ACID block is a DictionaryBlock

### PR DESCRIPTION
**Pull request by djsstarburst**
This PR consists of two commits:

Before the first commit, `Block.getChildren()` was used to take apart the
DELETE ACID rowId block.  That is always the wrong thing to do.
Fixed by using `ColumnarRow` to access the elements of the ACID rowId block.

This first commit also adds a comment warning developers not to use
`Block.getChildren()` so others don't make the same mistake of
calling the method.

The second commit replaces use of Block.getChildren in
HiveUpdateProcessor with ColumnarRow methods.

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java [hunk 1]
```diff
@@ -116,28 +117,31 @@ public HiveUpdatablePageSource(
     @Override
     public void deleteRows(Block rowIds)
     {
-        List<Block> blocks = rowIds.getChildren();
-        checkArgument(blocks.size() == 3, "The rowId block for DELETE should have 3 children, but has %s", blocks.size());
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java:120]: Should this check survive in the form of `columnarRow.getFieldCount()==3` inside `deleteRowsInternal`?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java [hunk 2]
```diff
@@ -116,28 +117,31 @@ public HiveUpdatablePageSource(
...
         deleteRowsInternal(rowIds);
     }
 
     private void deleteRowsInternal(Block rowIds)
     {
         int positionCount = rowIds.getPositionCount();
-        List<Block> blocks = rowIds.getChildren();
+        ColumnarRow columnarRow = ColumnarRow.toColumnarRow(rowIds);
+        for (int position = 0; position < positionCount; position++) {
+            // Throw an exception if there are any null rows
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java:128]: redundant 
```suggestion
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java [hunk 3]
```diff
@@ -116,28 +117,31 @@ public HiveUpdatablePageSource(
...
+            checkArgument(!columnarRow.isNull(position), "In the deleteRows block, found null position %s", position);
```
Comment (findepi) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java:129]: `deleteRows` is not a variable or anything else known

```suggestion
            checkArgument(!columnarRow.isNull(position), "In the delete rowIds, found null row at position %s", position);
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 4]
```diff
@@ -832,6 +832,28 @@ public void testDeleteAllRowsInPartition()
         });
     }
 
+    @Test(groups = HIVE_TRANSACTIONAL, timeOut = TEST_TIMEOUT)
+    public void testDeleteAllRowsUnpartitioned()
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:836]: i think the important aspect is "delete after delete".

assuming it's correct, let's reflect this in a test name: `testDeleteAfterDelete`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 5]
```diff
@@ -832,6 +832,28 @@ public void testDeleteAllRowsInPartition()
...
+            log.info("About to delete one row");
+            onTrino().executeQuery(format("DELETE FROM %s WHERE id = 2", tableName));
+
+            log.info("About to verify");
+            verifySelectForTrinoAndHive("SELECT * FROM " + tableName, "TRUE", row(1), row(3));
+
+            log.info("About to delete all rows");
+            onTrino().executeQuery("DELETE FROM " + tableName);
+
+            verifySelectForTrinoAndHive("SELECT COUNT(*) FROM " + tableName, "TRUE", row(0));
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:853]: nit: lowercase `count`, `true`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 6]
```diff
@@ -832,6 +832,28 @@ public void testDeleteAllRowsInPartition()
...
+            onTrino().executeQuery(format("CREATE TABLE %s (id INT) WITH (transactional = true)", tableName));
+
+            log.info("About to insert");
+            onTrino().executeQuery(format("INSERT INTO %s VALUES (1), (2), (3)", tableName));
+
+            log.info("About to delete one row");
+            onTrino().executeQuery(format("DELETE FROM %s WHERE id = 2", tableName));
+
+            log.info("About to verify");
+            verifySelectForTrinoAndHive("SELECT * FROM " + tableName, "TRUE", row(1), row(3));
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:848]: nit: lowercase `true`

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 7]
```diff
@@ -832,6 +832,28 @@ public void testDeleteAllRowsInPartition()
...
+
+            log.info("About to delete all rows");
+            onTrino().executeQuery("DELETE FROM " + tableName);
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:851]: Was the bug triggered for `DELETE` without any predicate, or for a delete with a predicate too?

we should test both cases

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java (hunk 6 above)
Comment (findepi) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:848]: verification of this stage is a bit redundant (we already have some tests for delete), especially if we run this on Hive too. remove, or keep only on Trino.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java [hunk 8]
```diff
@@ -832,6 +832,28 @@ public void testDeleteAllRowsInPartition()
         });
     }
 
+    @Test(groups = HIVE_TRANSACTIONAL, timeOut = TEST_TIMEOUT)
+    public void testDeleteAllRowsUnpartitioned()
+    {
+        withTemporaryTable("delete_all_rows", true, false, NONE, tableName -> {
+            onTrino().executeQuery(format("CREATE TABLE %s (id INT) WITH (transactional = true)", tableName));
+
+            log.info("About to insert");
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:841]: I agree with @martint on https://github.com/trinodb/trino/pull/9295/files#r714984993

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
cc @losipiuk

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java (hunk 1 above)
Comment (djsstarburst) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java:120]: Good idea, added.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java (hunk 2 above)
Comment (djsstarburst) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java]: Removed.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java (hunk 3 above)
Comment (djsstarburst) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveUpdatablePageSource.java]: Replaced.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java (hunk 4 above)
Comment (djsstarburst) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java]: Good idea; renamed.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java (hunk 5 above)
Comment (djsstarburst) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java]: Changed.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java (hunk 6 above)
Comment (djsstarburst) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java]: Changed to test only on Trino.

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java (hunk 7 above)
Comment (djsstarburst) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java:847]: Good idea.  I added a test that differs in this way:

```
            // A predicate sufficient to fool statistics-based optimization
            onTrino().executeQuery(format("DELETE FROM %s WHERE id != 2" + tableName));
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java (hunk 8 above)
Comment (djsstarburst) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestHiveTransactionalTable.java]: All logging removed from this method.

**Comment by djsstarburst** [comment]
Thanks for the prompt, close review, @findepi.  I force pushed an update making all your suggested changes.

BTW, separating the WHERE clause out as a separate argument to the `TestHiveTransactionalTable.verifySelect*` methods, which I introduced, was a dumb idea.  I will submit a no-semantic-change PR to remove them.

File: core/trino-spi/src/main/java/io/trino/spi/block/Block.java [hunk 9]
```diff
@@ -311,7 +311,9 @@ default Block getLoadedBlock()
     }
 
     /**
-     * Gets the direct child blocks of this block.
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/block/Block.java:314]: @dain should review this

**Comment by findepi** [comment]
CI https://github.com/trinodb/trino/issues/8432

**Comment by findepi** [comment]
Merged as 87785b424f570c04b67361b1b27256be5c0e437d, thanks!

File: core/trino-spi/src/main/java/io/trino/spi/block/Block.java (hunk 9 above)
Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/block/Block.java:314]: backed out

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/block/Block.java:314]: @djsstarburst can you please create a new PR with this change?

**Pull request closed without merging**